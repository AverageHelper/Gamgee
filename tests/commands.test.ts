import requireEnv from "./requireEnv";
import path from "path";
import { testerClient, isClientLoggedIn, commandResponseInSameChannel } from "./discordUtils";
import { fsUnlink } from "./fsUtils";

const COMMAND_PREFIX = requireEnv("BOT_PREFIX");
const TESTER_ID = requireEnv("CORDE_BOT_ID");
const QUEUE_CHANNEL_ID = requireEnv("QUEUE_CHANNEL_ID");

async function deleteTestDatabase() {
  const dbDir = path.resolve(__dirname, "./db/db.sqlite");
  await fsUnlink(dbDir);
}

describe("Command", () => {
  const PERMISSION_ERROR_RESPONSE = "YOU SHALL NOT PAAAAAASS!\nOr, y'know, something like that...";

  beforeAll(async () => {
    await deleteTestDatabase();
  });

  afterAll(async () => {
    // Log out of Discord
    if (isClientLoggedIn) {
      const client = await testerClient();
      client.destroy();
    }
  });

  describe("video", () => {
    const url = "https://youtu.be/dQw4w9WgXcQ";
    const info = `<@${TESTER_ID}>, Rick Astley - Never Gonna Give You Up (Video): (3 minutes, 32 seconds)`;

    test("returns the title and duration of a song with normal spacing", async () => {
      const response = await commandResponseInSameChannel(`video ${url}`);
      expect(response?.content).toBe(info);
    });

    test("returns the title and duration of a song with suboptimal spacing", async () => {
      const response = await commandResponseInSameChannel(`video ${url}`);
      expect(response?.content).toBe(info);
    });
  });

  describe("sr", () => {
    test("returns the queue instructional text", async () => {
      const response = await commandResponseInSameChannel("sr info");
      expect(response?.content).toBe(
        `To submit a song, type \`${COMMAND_PREFIX}sr <link>\`.
For example: \`${COMMAND_PREFIX}sr https://youtu.be/dQw4w9WgXcQ\`
I will respond with a text verification indicating your song has joined the queue!`
      );
    });

    test("yells at the tester for trying to set up a queue", async () => {
      let response = await commandResponseInSameChannel("sr setup");
      expect(response?.content).toBe(PERMISSION_ERROR_RESPONSE);

      response = await commandResponseInSameChannel(`sr setup <#${QUEUE_CHANNEL_ID}>`);
      expect(response?.content).toBe(PERMISSION_ERROR_RESPONSE);
    });

    test("yells at the tester for trying to open a queue", async () => {
      const response = await commandResponseInSameChannel("sr open");
      expect(response?.content).toBe(PERMISSION_ERROR_RESPONSE);
    });

    test("yells at the tester for trying to close the queue", async () => {
      const response = await commandResponseInSameChannel("sr close");
      expect(response?.content).toBe(PERMISSION_ERROR_RESPONSE);
    });

    test("yells at the tester for trying to set limits on the queue", async () => {
      const response = await commandResponseInSameChannel("sr limit");
      expect(response?.content).toBe(PERMISSION_ERROR_RESPONSE);
    });

    test("yells at the tester for trying to see queue statistics", async () => {
      const response = await commandResponseInSameChannel("sr stats");
      expect(response?.content).toBe(PERMISSION_ERROR_RESPONSE);
    });

    test("yells at the tester for trying to restart the queue", async () => {
      const response = await commandResponseInSameChannel("sr restart");
      expect(response?.content).toBe(PERMISSION_ERROR_RESPONSE);
    });

    // TODO: Add a test for sr <song link>
  });

  describe("help", () => {
    test("returns the help text", async () => {
      const response = await commandResponseInSameChannel("help");
      expect(response?.content).toMatchSnapshot();
    });
  });
});
