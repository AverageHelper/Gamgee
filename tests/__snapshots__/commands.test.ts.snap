// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Command help returns the help text 1`] = `
"Commands:
\`?config\` - Read and modify config options. *(Server owner only. No touch!)*
\`?ping\` - Ping my host server to check latency.
\`?sr <YouTube or SoundCloud link>\` - Submit a song to the queue.
    \`?sr info\` - Prints a handy message to let people know how to queue-up.
    \`?sr setup <channel name>\` - Set a channel as the 'queue' channel. *(Server owner only. No touch!)*
    \`?sr open\` - Start accepting song requests to the queue.
    \`?sr close\` - Stop accepting song requests to the queue.
    \`?sr limit <entry-duration|cooldown|count>\` - Sets a limit value on the queue. (Time in seconds, where applicable)
    \`?sr stats\` - Reports statistics on the current queue.
    \`?sr restart\` - Empties the queue and starts a fresh queue session.
\`?t\` - Start typing :wink:
\`?video <YouTube or SoundCloud link>\` - Puts the video title and duration in chat."
`;
